nomad/structs/structs.go

```diff
@@ -7632,6 +7634,39 @@ func (a *Allocation) CopySkipJob() *Allocation {
 	return a.copyImpl(false)
 }
 
+func (a *Allocation) Canonicalize() {
+	if a.AllocatedResources == nil && a.TaskResources != nil {
+		ar := AllocatedResources{}
+		ar.Tasks = toAllocatedResources(a.TaskResources)
+
+		if a.SharedResources != nil {
+			ar.Shared.DiskMB = int64(a.SharedResources.DiskMB)
+			ar.Shared.Networks = a.SharedResources.Networks.Copy()
+		}
+
+		a.AllocatedResources = &ar
+	}
+
+	// TODO: Investigate if we should canonicalize the job
+	// it may be out of sync with respect to the original job
+	// a.Job.Canonicalize()
```

Comment:
```suggestion
	a.Job.Canonicalize()
```

I think this is the right thing to do:
* On clients they may have an old version of a job persisted that may not match the expectations of the code running in the agent.
* On servers I actually don't remember off the top of my head if Jobs are persisted on Allocs (are allocs normalized only in raft messages on the wire, or in state store too :thinking:  ?). **Luckily this doesn't matter!** Job.Canonicalize handles job being `nil` and Jobs are already canonicalized on statestore restore.

So I think we might waste a couple CPU instructions, but it seems necessary on clients at least.